pennylane/grouping/utils.py

```diff
@@ -0,0 +1,589 @@
+    else:
+        operations_zip = zip(pauli_wires, pauli_word.name)
+
+    for wire, name in operations_zip:
+
+        if name == "PauliX":
+            binary_pauli[wire] = 1
+
+        elif name == "PauliY":
+            binary_pauli[wire] = 1
+            binary_pauli[n_qubits + wire] = 1
+
+        elif name == "PauliZ":
+            binary_pauli[n_qubits + wire] = 1
+
+    return binary_pauli
+
+
+def binary_to_pauli(binary_vector, wire_map=None):  # pylint: disable=too-many-branches
+    """Converts a binary vector of even dimension to an Observable instance.
+
+    This functions follows convention that the first half of binary vector components specify
+    PauliX placements while the last half specify PauliZ placements.
+
+    **Usage example:**
+
+    If `wire_map` is unspecified, the Pauli operations follow the same enumerations as the vector
+    components, i.e. the `i` and `N+i` components specify the Pauli operation on wire `i`,
+
+    >>> binary_to_pauli([0,1,1,0,1,0])
+    Tensor(PauliY(wires=[1]), PauliX(wires=[2]))
+
+    An arbitrary labelling can be assigned by using `wire_map`
+    >>> wire_map = {Wires('a'): 0, Wires('b'): 1, Wires('c'): 2}
+    >>> binary_to_pauli([0,1,1,0,1,0], wire_map=wire_map)
+    Tensor(PauliY(wires=['b']), PauliX(wires=['c']))
+
+    Note that the values of `wire_map`, if specified, must be 0,1,..., N, where N is the dimension
+    of the vector divided by two, i.e. `list(wire_map.values())` must be
+    `list(range(len(binary_vector)/2))`.
+
+    Args:
+        binary_vector (Union[list, tuple, array]): binary vector of even dimension representing a
+            unique Pauli word.
+
+    Keyword args:
+        wire_map (dict): dictionary containing all wire labels used in the Pauli word as keys, and
+                         unique integer labels as their values
+
+    Returns:
+        Tensor(Union[Identity, PauliX, PauliY, PauliZ]): the Pauli word corresponding to the input
+            binary vector. Note that if a zero-vector is input, then the resulting Pauli word will
+            be an `Identity` instance.
+
+    Raises:
+        TypeError: if length of binary vector is not even, or if vector does not have strictly
+            binary components.
+
+    """
+
+    if isinstance(binary_vector, (list, tuple)):
+        binary_vector = np.asarray(binary_vector)
+
+    if len(binary_vector) % 2 != 0:
+        raise ValueError(
+            "Length of binary_vector must be even, instead got vector of shape {}.".format(
+                np.shape(binary_vector)
+            )
+        )
+
+    if not np.array_equal(binary_vector, binary_vector.astype(bool)):
+        raise ValueError(
+            "Input vector must have strictly binary components, instead got {}.".format(
+                binary_vector
+            )
+        )
+
+    n_qubits = len(binary_vector) // 2
+
+    if wire_map is not None:
+        if list(wire_map.values()) != list(range(n_qubits)):
```

Comment:
```suggestion
        if set(wire_map.values()) != set(range(n_qubits)):
```
Seems to have a problem if values were given in a non-ascending order:
```python
from pennylane.wires import Wires

wire_map = {Wires('a'): 1, Wires('b'): 0, Wires('c'): 2}
binary_to_pauli([0,1,1,0,1,0], wire_map=wire_map)
```
outputs
```    392             raise ValueError(
    393                 "The values of wire_map must be integers 0 to N, for 2N-dimensional binary vector."
--> 394                 " Instead got wire_map values: {}".format(wire_map.values())
    395             )
    396         label_map = {explicit_index: wire_label for wire_label, explicit_index in wire_map.items()}

ValueError: The values of wire_map must be integers 0 to N, for 2N-dimensional binary vector. Instead got wire_map values: dict_values([1, 0, 2])
```

Reply:
Great catch, thanks!